File: androidApp/src/main/java/com/walletline/android/presentation/screens/auth/pattern/make_pattern/MakePatternViewModel.kt
package com.walletline.android.presentation.screens.auth.pattern.make_pattern

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.walletline.domain.model.BiometricType
import com.walletline.domain.model.PatternValidationMessage
import com.walletline.domain.use_case.auth.AuthUseCase
import com.walletline.presentation.screens.auth.pattern.make_pattern.MakePatternState
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class MakePatternViewModel(
    private val authUseCase: AuthUseCase,
) : ViewModel(), MakePatternContract {


    private val _state = MutableStateFlow(MakePatternState())
    override val state: StateFlow<MakePatternState> = _state.asStateFlow()

    private val effectChannel = Channel<MakePatternContract.Effect>()
    override val effect: Flow<MakePatternContract.Effect> = effectChannel.receiveAsFlow()

    private val clickState = Channel<MakePatternContract.Event>()

    init {
        clickState
            .receiveAsFlow()
            .debounce(400)
            .onEach { event ->
                when (event) {
                    MakePatternContract.Event.ContinueClicked -> effectChannel.send(
                        MakePatternContract.Effect.NavigateToConfirmPattern
                    )

                    MakePatternContract.Event.FingerPrintClicked -> effectChannel.send(
                        MakePatternContract.Effect.ShowBiometricPrompt
                    )

                    MakePatternContract.Event.ConfirmPatternClicked -> confirmPattern()
                    MakePatternContract.Event.RetryPatternClicked -> effectChannel.send(
                        MakePatternContract.Effect.RetryPattern
                    )

                    MakePatternContract.Event.SkipPatternClicked -> handleSkipPattern()

                    is MakePatternContract.Event.FingerFaceSuccessful -> Unit
                    is MakePatternContract.Event.FingerFaceUnSuccessful -> Unit
                    is MakePatternContract.Event.PatternChanged -> Unit
                    is MakePatternContract.Event.ConfirmPatternChange -> Unit
                }
            }
            .launchIn(viewModelScope)
    }


    override fun onEvent(event: MakePatternContract.Event) {
        when (event) {
            is MakePatternContract.Event.PatternChanged -> updatePatternAndButtonState(event)
            is MakePatternContract.Event.ConfirmPatternChange -> updatePatternAndButtonState(event)
            MakePatternContract.Event.ContinueClicked,
            MakePatternContract.Event.FingerPrintClicked,
            MakePatternContract.Event.ConfirmPatternClicked,
            MakePatternContract.Event.RetryPatternClicked,
            MakePatternContract.Event.SkipPatternClicked,
            -> viewModelScope.launch { clickState.send(event) }

            is MakePatternContract.Event.FingerFaceSuccessful -> handleBiometricSuccess(event.type)
            is MakePatternContract.Event.FingerFaceUnSuccessful -> TODO()
        }
    }

    private fun handleSkipPattern() {
        viewModelScope.launch {
            authUseCase.setAdmission.execute(pattern = "", isFingerprint = false)
            effectChannel.send(MakePatternContract.Effect.SkipMakePattern)
        }
    }

    private fun confirmPattern() {
        if (state.value.pattern != state.value.confirmPattern) {
            effectChannel.trySend(
                element = MakePatternContract.Effect.ConfirmPatternUnSuccessful(
                    message = PatternValidationMessage.NotSame
                )
            )
            return
        }
        viewModelScope.launch {
            authUseCase.setAdmission.execute(
                pattern = state.value.confirmPattern,
                isFingerprint = false
            )
            effectChannel.send(MakePatternContract.Effect.ConfirmPatternSuccessful)
        }
    }

    private fun handleBiometricSuccess(type: BiometricType) {
        viewModelScope.launch {
            authUseCase.setAdmission.execute(pattern = "", isFingerprint = true)
            effectChannel.send(MakePatternContract.Effect.BiometricSuccess)
        }
    }

    private fun updatePatternAndButtonState(event: MakePatternContract.Event) {
        when (event) {
            is MakePatternContract.Event.PatternChanged -> _state.update { state ->
                state.copy(
                    pattern = event.pattern,
                    isScrollEnabled = event.isScrollEnabled,
                    isContinueButtonEnable = event.pattern.length >= 4
                )
            }

            is MakePatternContract.Event.ConfirmPatternChange -> _state.update { state ->
                state.copy(
                    confirmPattern = event.pattern,
                    isScrollEnabled = event.isScrollEnabled,
                    isContinueButtonEnable = event.pattern.length >= 4
                )
            }

            else -> Unit
        }

    }

}

File: iosApp/iosApp/presentation/screens/auth_screens/pattern_screen/make_pattern/MakePatternViewModel.swift
//
//  MakePatternViewModel.swift
//  iosApp
//
//  Created by Mostafa Hosseini on 5/27/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Combine
import Foundation
import KMPNativeCoroutinesCombine
import shared

@MainActor
class MakePatternViewModel: ObservableObject {
    private let authUseCase = KoinHelper().authUseCase
    private var cancellables = Set<AnyCancellable>()

    @Published private(set) var state = MakePatternState()
    private let effectSubject = PassthroughSubject<MakePatternEffect, Never>()

    func effect() -> AnyPublisher<MakePatternEffect, Never> {
        return effectSubject.eraseToAnyPublisher()
    }

    func onEvent(_ event: MakePatternEvent) {
        switch event {
        case .PatternChanged(let pattern, let isScrollEnabled):
            updateState(pattern: pattern, isScrollEnabled: isScrollEnabled, isContinueButtonEnable: pattern.count >= 4)
        case .ConfirmPatternChange(pattern: let pattern, isScrollEnabled: let isScrollEnabled):
            updateState(confirmPattern: pattern, isScrollEnabled: isScrollEnabled, isContinueButtonEnable: pattern.count >= 4)
        case .SkipPatternClicked:
            handleSkipPattern()
        case .ContinueClicked:
            effectSubject.send(.NavigateToConfirmPattern)
        case .FingerPrintClicked:
            effectSubject.send(.ShowBiometricPrompt)
        case .ConfirmPatternClicked:
            confirmPattern()
        case .RetryPatternClicked:
            effectSubject.send(.RetryPattern)
        case .FingerFaceSuccessful(type: _):
            handleSuccessFingerFace()
        case .FingerFaceUnSuccessful(error: let error):
            handleUnsuccessfulFingerFace(error)
        }
    }

    private func handleUnsuccessfulFingerFace(_ error: BiometricError) {}

    private func handleSuccessFingerFace() {
        let setAdmission = createFuture(for: authUseCase.setAdmission.execute(pattern: "", isFingerprint: true))
        setAdmission.sink(receiveCompletion: { _ in }, receiveValue: { _ in
            self.effectSubject.send(.BiometricSuccess)
        })
        .store(in: &cancellables)
    }

    private func handleSkipPattern() {
        let setAdmission = createFuture(for: authUseCase.setAdmission.execute(pattern: "", isFingerprint: false))
        setAdmission.sink(receiveCompletion: { _ in

        }, receiveValue: { _ in
            self.effectSubject.send(.SkipMakePattern)
        }).store(in: &cancellables)
    }

    private func confirmPattern() {
        if state.pattern != state.confirmPattern {
            effectSubject.send(.ConfirmPatternUnSuccessful(message: PatternValidationMessageNotSame()))
            return
        }

        let setAdmission = createFuture(for: authUseCase.setAdmission.execute(pattern: state.pattern, isFingerprint: false))
        setAdmission.sink(receiveCompletion: { _ in

        }, receiveValue: { _ in
            self.effectSubject.send(.ConfirmPatternSuccessful)
        })
        .store(in: &cancellables)
    }

    private func updateState(
        pattern: String? = nil,
        confirmPattern: String? = nil,
        patternError: PatternValidationMessage? = nil,
        isScrollEnabled: Bool? = nil,
        isContinueButtonEnable: Bool? = nil,
        isConfirmButtonEnable: Bool? = nil
    ) {
        guard let currentState = state.copy() as? MakePatternState else { return }
        DispatchQueue.main.async { [self] in
            state = state.doCopy(
                pattern: pattern ?? currentState.pattern,
                confirmPattern: confirmPattern ?? currentState.confirmPattern,
                patternError: patternError ?? currentState.patternError,
                isScrollEnabled: isScrollEnabled ?? currentState.isScrollEnabled,
                isContinueButtonEnable: isContinueButtonEnable ?? currentState.isContinueButtonEnable,
                isConfirmButtonEnable: isConfirmButtonEnable ?? currentState.isConfirmButtonEnable
            )
        }
    }

    private func cancelAsyncTasks() {
        cancellables.forEach { cancellable in
            cancellable.cancel()
        }
    }

    private func doNothing() {}
}

enum MakePatternEvent {
    case PatternChanged(pattern: String, isScrollEnabled: Bool)
    case ConfirmPatternChange(pattern: String, isScrollEnabled: Bool)
    case SkipPatternClicked
    case ContinueClicked
    case FingerPrintClicked
    case ConfirmPatternClicked
    case RetryPatternClicked
    case FingerFaceSuccessful(type: BiometricType)
    case FingerFaceUnSuccessful(error: BiometricError)
}

enum MakePatternEffect {
    case NavigateToConfirmPattern
    case ConfirmPatternSuccessful
    case BiometricSuccess
    case ConfirmPatternUnSuccessful(message: PatternValidationMessage)
    case RetryPattern
    case ShowBiometricPrompt
    case SkipMakePattern
}
